File: 2023/day10/go/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// Direction deltas: N=(-1,0), S=(1,0), E=(0,1), W=(0,-1)
var pipeConnections = map[byte][][2]int{
	'|': {{-1, 0}, {1, 0}},  // N, S
	'-': {{0, -1}, {0, 1}},  // W, E
	'L': {{-1, 0}, {0, 1}},  // N, E
	'J': {{-1, 0}, {0, -1}}, // N, W
	'7': {{1, 0}, {0, -1}},  // S, W
	'F': {{1, 0}, {0, 1}},   // S, E
}

type Point struct {
	r, c int
}

func findStart(grid []string) Point {
	for r, row := range grid {
		for c := 0; c < len(row); c++ {
			if row[c] == 'S' {
				return Point{r, c}
			}
		}
	}
	return Point{-1, -1}
}

func getNeighbors(grid []string, pos Point) []Point {
	r, c := pos.r, pos.c
	rows, cols := len(grid), len(grid[0])
	ch := grid[r][c]

	var neighbors []Point

	if ch == 'S' {
		// S can connect to any adjacent pipe that connects back to it
		directions := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
		for _, d := range directions {
			nr, nc := r+d[0], c+d[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				adjCh := grid[nr][nc]
				if dirs, ok := pipeConnections[adjCh]; ok {
					// Check if adjacent pipe connects back to S
					for _, adjD := range dirs {
						if nr+adjD[0] == r && nc+adjD[1] == c {
							neighbors = append(neighbors, Point{nr, nc})
							break
						}
					}
				}
			}
		}
	} else if dirs, ok := pipeConnections[ch]; ok {
		for _, d := range dirs {
			nr, nc := r+d[0], c+d[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				neighbors = append(neighbors, Point{nr, nc})
			}
		}
	}

	return neighbors
}

func findLoop(grid []string, start Point) map[Point]int {
	distances := make(map[Point]int)
	distances[start] = 0
	queue := []Point{start}

	for len(queue) > 0 {
		pos := queue[0]
		queue = queue[1:]

		for _, neighbor := range getNeighbors(grid, pos) {
			if _, exists := distances[neighbor]; !exists {
				distances[neighbor] = distances[pos] + 1
				queue = append(queue, neighbor)
			}
		}
	}

	return distances
}

func determineStartPipe(grid []string, start Point, loopPositions map[Point]int) byte {
	r, c := start.r, start.c
	rows, cols := len(grid), len(grid[0])

	connections := make(map[[2]int]bool)
	directions := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for _, d := range directions {
		nr, nc := r+d[0], c+d[1]
		neighbor := Point{nr, nc}
		if _, inLoop := loopPositions[neighbor]; inLoop {
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				adjCh := grid[nr][nc]
				if dirs, ok := pipeConnections[adjCh]; ok {
					for _, adjD := range dirs {
						if nr+adjD[0] == r && nc+adjD[1] == c {
							connections[d] = true
							break
						}
					}
				}
			}
		}
	}

	for pipe, dirs := range pipeConnections {
		if len(dirs) == len(connections) {
			match := true
			for _, d := range dirs {
				if !connections[d] {
					match = false
					break
				}
			}
			if match {
				return pipe
			}
		}
	}

	return 'S'
}

func part1(grid []string) int {
	start := findStart(grid)
	distances := findLoop(grid, start)

	maxDist := 0
	for _, d := range distances {
		if d > maxDist {
			maxDist = d
		}
	}
	return maxDist
}

func part2(grid []string) int {
	start := findStart(grid)
	distances := findLoop(grid, start)

	// Replace S with its actual pipe type
	startPipe := determineStartPipe(grid, start, distances)
	modifiedGrid := make([][]byte, len(grid))
	for i, row := range grid {
		modifiedGrid[i] = []byte(row)
	}
	modifiedGrid[start.r][start.c] = startPipe

	rows, cols := len(modifiedGrid), len(modifiedGrid[0])
	enclosed := 0

	for r := 0; r < rows; r++ {
		inside := false
		for c := 0; c < cols; c++ {
			pos := Point{r, c}
			if _, inLoop := distances[pos]; inLoop {
				ch := modifiedGrid[r][c]
				// Count pipes that have a north connection: |, L, J
				if ch == '|' || ch == 'L' || ch == 'J' {
					inside = !inside
				}
			} else {
				if inside {
					enclosed++
				}
			}
		}
	}

	return enclosed
}

func main() {
	_, currentFile, _, _ := runtime.Caller(0)
	dir := filepath.Dir(currentFile)
	inputPath := filepath.Join(dir, "..", "input.txt")

	data, err := os.ReadFile(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 2: %d\n", part2(lines))
}

File: 2023/day10/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

# Read input
my $script_dir = dirname(__FILE__);
my $input_path = File::Spec->catfile($script_dir, '..', 'input.txt');
open my $fh, '<', $input_path or die "Cannot open $input_path: $!";
my @lines = <$fh>;
close $fh;
chomp @lines;

# Pipe connections: each pipe connects to certain directions
# Directions: N=(-1,0), S=(1,0), E=(0,1), W=(0,-1)
my %PIPE_CONNECTIONS = (
    '|' => [[-1, 0], [1, 0]],    # N, S
    '-' => [[0, -1], [0, 1]],    # W, E
    'L' => [[-1, 0], [0, 1]],    # N, E
    'J' => [[-1, 0], [0, -1]],   # N, W
    '7' => [[1, 0], [0, -1]],    # S, W
    'F' => [[1, 0], [0, 1]],     # S, E
);

sub find_start {
    my ($grid) = @_;
    for my $r (0 .. $#$grid) {
        my @chars = split //, $grid->[$r];
        for my $c (0 .. $#chars) {
            if ($chars[$c] eq 'S') {
                return ($r, $c);
            }
        }
    }
    return (undef, undef);
}

sub get_neighbors {
    my ($grid, $r, $c) = @_;
    my $rows = scalar @$grid;
    my $cols = length $grid->[0];
    my $ch = substr($grid->[$r], $c, 1);
    my @neighbors;

    if ($ch eq 'S') {
        # S can connect to any adjacent pipe that connects back to it
        for my $dir ([-1, 0], [1, 0], [0, -1], [0, 1]) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;
            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
                my $adj_ch = substr($grid->[$nr], $nc, 1);
                if (exists $PIPE_CONNECTIONS{$adj_ch}) {
                    for my $adj_dir (@{$PIPE_CONNECTIONS{$adj_ch}}) {
                        my ($adj_dr, $adj_dc) = @$adj_dir;
                        if ($nr + $adj_dr == $r && $nc + $adj_dc == $c) {
                            push @neighbors, [$nr, $nc];
                            last;
                        }
                    }
                }
            }
        }
    } elsif (exists $PIPE_CONNECTIONS{$ch}) {
        for my $dir (@{$PIPE_CONNECTIONS{$ch}}) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;
            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
                push @neighbors, [$nr, $nc];
            }
        }
    }

    return @neighbors;
}

sub find_loop {
    my ($grid, $start_r, $start_c) = @_;
    my %distances;
    my $start_key = "$start_r,$start_c";
    $distances{$start_key} = 0;
    my @queue = ([$start_r, $start_c]);

    while (@queue) {
        my $pos = shift @queue;
        my ($r, $c) = @$pos;
        my $curr_key = "$r,$c";

        for my $neighbor (get_neighbors($grid, $r, $c)) {
            my ($nr, $nc) = @$neighbor;
            my $neighbor_key = "$nr,$nc";
            if (!exists $distances{$neighbor_key}) {
                $distances{$neighbor_key} = $distances{$curr_key} + 1;
                push @queue, [$nr, $nc];
            }
        }
    }

    return \%distances;
}

sub determine_start_pipe {
    my ($grid, $start_r, $start_c, $loop_positions) = @_;
    my $rows = scalar @$grid;
    my $cols = length $grid->[0];
    my %connections;

    for my $dir ([-1, 0], [1, 0], [0, -1], [0, 1]) {
        my ($dr, $dc) = @$dir;
        my $nr = $start_r + $dr;
        my $nc = $start_c + $dc;
        my $neighbor_key = "$nr,$nc";

        if (exists $loop_positions->{$neighbor_key}) {
            my $adj_ch = substr($grid->[$nr], $nc, 1);
            if (exists $PIPE_CONNECTIONS{$adj_ch}) {
                for my $adj_dir (@{$PIPE_CONNECTIONS{$adj_ch}}) {
                    my ($adj_dr, $adj_dc) = @$adj_dir;
                    if ($nr + $adj_dr == $start_r && $nc + $adj_dc == $start_c) {
                        $connections{"$dr,$dc"} = 1;
                        last;
                    }
                }
            }
        }
    }

    for my $pipe (keys %PIPE_CONNECTIONS) {
        my @dirs = @{$PIPE_CONNECTIONS{$pipe}};
        my %pipe_dirs;
        for my $d (@dirs) {
            $pipe_dirs{"$d->[0],$d->[1]"} = 1;
        }

        my $match = 1;
        if (scalar(keys %connections) != scalar(keys %pipe_dirs)) {
            $match = 0;
        } else {
            for my $k (keys %connections) {
                if (!exists $pipe_dirs{$k}) {
                    $match = 0;
                    last;
                }
            }
        }

        if ($match) {
            return $pipe;
        }
    }

    return 'S';
}

sub part1 {
    my ($start_r, $start_c) = find_start(\@lines);
    my $distances = find_loop(\@lines, $start_r, $start_c);
    my $max_dist = 0;
    for my $d (values %$distances) {
        $max_dist = $d if $d > $max_dist;
    }
    return $max_dist;
}

sub part2 {
    my ($start_r, $start_c) = find_start(\@lines);
    my $distances = find_loop(\@lines, $start_r, $start_c);
    my $loop_positions = $distances;

    # Determine what pipe S actually is
    my $start_pipe = determine_start_pipe(\@lines, $start_r, $start_c, $loop_positions);

    # Create a mutable copy of the grid
    my @grid = map { [split //, $_] } @lines;
    $grid[$start_r][$start_c] = $start_pipe;

    my $rows = scalar @grid;
    my $cols = scalar @{$grid[0]};
    my $enclosed = 0;

    for my $r (0 .. $rows - 1) {
        my $inside = 0;
        for my $c (0 .. $cols - 1) {
            my $key = "$r,$c";
            if (exists $loop_positions->{$key}) {
                my $ch = $grid[$r][$c];
                # Count vertical crossings using "north" rule
                # Count pipes that have a north connection: |, L, J
                if ($ch eq '|' || $ch eq 'L' || $ch eq 'J') {
                    $inside = !$inside;
                }
            } else {
                if ($inside) {
                    $enclosed++;
                }
            }
        }
    }

    return $enclosed;
}

print "Part 1: ", part1(), "\n";
print "Part 2: ", part2(), "\n";
